jolt-core/src/poly/commitment/bmmtv/poly_commit.rs

```diff
@@ -0,0 +1,411 @@
+use ark_std::rand::Rng;
+use rand_core::CryptoRng;
+
+pub struct BivariatePolynomial<F: Field> {
+    y_polynomials: Vec<UnivariatePolynomial<F>>,
+}
+
+impl<F: Field> BivariatePolynomial<F>
+where
+    F: JoltField,
+{
+    pub fn evaluate(&self, point: &(F, F)) -> F {
+        let (x, y) = point;
+        let mut point_x_powers = vec![];
+        let mut cur = F::one();
+        for _ in 0..(self.y_polynomials.len()) {
+            point_x_powers.push(cur);
+            cur *= x;
+        }
+        point_x_powers
+            .iter()
+            .zip(&self.y_polynomials)
+            .map(|(x_power, y_polynomial)| *x_power * y_polynomial.evaluate(y))
+            .sum()
+    }
+}
+
+pub struct OpeningProof<P: Pairing> {
+    ip_proof: MippKProof<P>,
+    y_eval_comm: P::G1,
+    kzg_proof: P::G1,
+}
+
+pub struct BivariatePolynomialCommitment<P: Pairing, D>(MippK<P, D>);
+
+impl<P: Pairing, ProofTranscript: Transcript> BivariatePolynomialCommitment<P, ProofTranscript>
+where
+    P::ScalarField: JoltField,
+    P::G1: Icicle,
+    P::G2: Icicle,
+{
+    pub fn setup<R: Rng + CryptoRng>(
+        rng: &mut R,
+        x_degree: usize,
+        y_degree: usize,
+    ) -> Result<SRS<P>, Error> {
+        Ok(SRS::setup(rng, y_degree, 2 * x_degree))
+    }
+
+    pub fn commit(
+        kzg_srs: &KZGProverKey<P>,
+        bivariate_polynomial: &BivariatePolynomial<P::ScalarField>,
+    ) -> Result<(PairingOutput<P>, Vec<P::G1>), Error> {
+        assert!((kzg_srs.g2_powers().len() + 1) / 2 >= bivariate_polynomial.y_polynomials.len());
+        let ck = kzg_srs
+            .g2_powers()
+            .iter()
+            .map(|aff| aff.into_group())
+            .step_by(2)
+            .collect::<Vec<_>>();
+
+        // Create KZG commitments to Y polynomials
+        let y_polynomial_coms = bivariate_polynomial
+            .y_polynomials
+            .iter()
+            .chain([UnivariatePolynomial::zero()].iter().cycle())
+            .take(ck.len())
+            .map(|y_polynomial| {
+                UnivariateKZG::<P>::commit(kzg_srs, y_polynomial).map(|affine| affine.into_group())
+            })
+            .collect::<Result<Vec<P::G1>, _>>()?;
+
+        // Create AFGHO commitment to Y polynomial commitments
+        Ok((
+            AfghoCommitment::<P>::commit(&ck, &y_polynomial_coms)?,
+            y_polynomial_coms,
+        ))
+    }
+
+    pub fn open(
+        kzg_srs: &KZGProverKey<P>,
+        bivariate_polynomial: &BivariatePolynomial<P::ScalarField>,
+        y_polynomial_comms: Vec<P::G1>,
+        point: &(P::ScalarField, P::ScalarField),
+        transcript: &mut ProofTranscript,
+    ) -> Result<OpeningProof<P>, Error> {
+        let (x, y) = point;
+        let commitment_key_len = (kzg_srs.g2_powers().len() + 1) / 2;
+        assert!(commitment_key_len >= bivariate_polynomial.y_polynomials.len());
+
+        let precomp_time = start_timer!(|| "Computing coefficients and KZG commitment");
```

Comment:
Instead of `start_timer!` and `end_timer!` we use the `tracing::instrument` decorator for functions and `tracing::span!` for inline profiling